### ISSUE TITLE
FTPFileListProtocol does not support files with spaces in them

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4986]
|<img alt="jrydberg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jrydberg reported|
|-|-|
|Trac ID|trac#4986|
|Type|defect|
|Created|2011-03-24 13:47:41Z|
|Branch|https://github.com/twisted/twisted/tree/ftp-filename-spaces-4986|

Have a directory with a few files with spaces in their name, and they do not show up in my listing when I use twisted.protocols.ftp.FTPFileListProtocol.

```
from twisted.protocols.ftp import FTPFileListProtocol
p = FTPFileListProtocol()
data1 = "-rwxr-xr-x    1 1003     1003     186352836 Mar 17 09:55 Big Buck Bunny_2436000.ismv"
data2 = "-rwxr-xr-x    1 1003     1003     186352836 Mar 17 09:55 Big_Buck_Bunny_2436000.ismv"
p.lineReceived(data1)
p.lineReceived(data2)
assert len(p.files) == 2
```

Twisted version 10.2

Attachments:

* [4986-file-with-space-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/80c/80ca9d381c26ceda17a0e478bb05f4ec8d1a097a/b34e021af1492869529141628e9e5c1dc55115ef.diff) (1933 bytes) - added by adiroiban on 2012-06-18 08:55:37Z - 
* [4986-file-with-space-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/80c/80ca9d381c26ceda17a0e478bb05f4ec8d1a097a/22d910ddd64ee49200152294e295948e5e3ea45c.diff) (2575 bytes) - added by adiroiban on 2012-06-18 09:09:21Z - 
* [4986-file-with-space-3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/80c/80ca9d381c26ceda17a0e478bb05f4ec8d1a097a/eb743046e0f015d1ce41967c809c41b2740bb7c9.diff) (3804 bytes) - added by adiroiban on 2012-07-20 13:20:57Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4986 4986
type__defect defect
reporter__jrydberg jrydberg
priority__normal normal
milestone__ 
branch__branches_ftp_filename_spaces_4986 branches/ftp-filename-spaces-4986
branch_author__tenth tenth
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1300974461000000 1300974461000000
changetime__1353248005000000 1353248005000000
version__None None
owner__tenth tenth
cc__itamarst cc__adiroiban
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban commented|
|-|-|

Hi,

I have attached a patch that attempts to fix this issue.

It replaces:
```
(?P<filename>([^ ]|\\ )*?)
```

with

```
(?P<filename>(.){1,}?)
```

I don't know why this expression was used: ``` ([^ ]|\\ )* ``` and why space characters were explicitly denied.

Maybe I am not understanding something right and my patch is not valid.

Please let me know if you spot any errors.

Adi

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|spiv set owner to @adiroiban|
|-|-|

Replying to [adiroiban](#note_2):
> I don't know why this expression was used: ``` ([^ ]|\\ )* ``` and why space characters were explicitly denied.

That appears to be intended to only match backslash-escaped spaces.  Note that parseDirectoryLine will unescape those.

I think it would be reasonable to attempt to accept unescaped spaces too, with appropriate care taken to ensure the existing unescaping functionality is preserved.  That means adding unit tests for it, as there don't appear to be any yet.

Aside from that, your patch looks sound.  I'm a bit hesitant to clutter the docstring with examples of various cases it handles, but I'd rather more documentation than less, so it's ok.

This patch will also need a news entry (i.e. add twisted/topfiles/4986.bug, as per http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles).

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban set owner to spiv|
|-|-|

Hi,

Many thanks for the review.

I have added the required news file together with a test for escaped strings.

Please let me know if other changes are required.
Adi

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_5"></a>|@tenth10th removed owner|
|-|-|

Hi adiroiban, just wanted to suggest that when a ticket is ready for re-review, you should probably leave it as unassigned, with the review keyword, to give it the best chance of being picked up by a reviewer.

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_6"></a>|@tenth10th commented|
|-|-|

(In [36274]) Branching to 'ftp-filename-spaces-4986'

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_7"></a>|@tenth10th commented|
|-|-|

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ftp-filename-spaces-4986) for patch 4986-file-with-space-3.diff

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_8"></a>|@tenth10th set owner to @tenth10th|
|-|-|

Hi adiroiban,

Unfortunately, the doc changes in twisted.protocols.ftp aren't properly formatted for epytext, and are currently causing a [build error](http://buildbot.twistedmatrix.com/builders/documentation/builds/2691/steps/api-documentation):

Please take a look at the [Tools For Development](http://twistedmatrix.com/trac/wiki/TwistedDevelopment#ToolsForDevelopment) page for more information on how to get pydoctor and epytext, which will allow you to test your doc changes.
And some documentation for epytext markup itself can be found [here](http://epydoc.sourceforge.net/manual-epytext.html).

Otherwise this patch looks good; I'm going to make a quick fix to the docs (to mark the examples starting at line 2874 as a literal block) and then land this branch.

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_9"></a>|@tenth10th commented|
|-|-|

(In [36378]) Made the examples a literal block: refs [#4986](https://github.com/twisted/twisted/issues/4986)

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_10"></a>|@tenth10th set status to `closed`|
|-|-|

(In [36380]) Merge ftp-filename-spaces-4986: FTPFileListProtocol does not support files with spaces in them

Author: adiroiban
Reviewer: spiv, tenth
Fixes: [#4986](https://github.com/twisted/twisted/issues/4986)

Fixes a bug wherein twisted.protocols.ftp.FTPFileListProtocol does not list filenames containing spaces.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban commented|
|-|-|

Thanks for the review and for the link to Development Tools.

Until now I was only using "trial" to validate the code.
I will try to compile epydoc before requesting a review.

Cheers,
Adi
